### ISSUE TITLE
[Bug]: npu_fused_infer_attention_score silently ignores atten_mask past KV position 512 (bool [1,1,1,S], sparse_mode=0)

### ISSUE THREAD

**Issue by Junbo2002** [vllm-project/vllm-ascend#15503]
### Summary

`torch_npu.npu_fused_infer_attention_score` with `sparse_mode=0` and a bool `atten_mask` of shape `[1, 1, 1, S]` **silently ignores the mask past KV position 512.** Up to and including `L=512` the result matches an fp32 reference; from `L=513` it is wrong, by up to 0.25 absolute, with no error and no warning.

512 is a hard constant. It does **not** track `block_size` and does not track the declared capacity:

| block_size | capacity | last KV length still correct |
| ---: | ---: | ---: |
| 16 | 2048 / 4096 | 512 |
| 32 | 2048 / 4096 | 512 |
| 64 | 2048 / 4096 | 512 |
| 128 | 2048 / 4096 | 512 |
| 256 | 2048 / 4096 | 512 |

A second under-ranked shape fails differently in the same call: `pse_shift` of shape `[1, 1, 1, S]` is wrong at almost every length. Giving `pse_shift` its full head dimension (`[1, N, 1, S]`) is correct everywhere, which is the workaround we ended up on.

A rank-2 bool mask `[1, S]` is *rejected* (`aclnnFusedInferAttentionScoreV3 561002`) — that is the operator behaving correctly. The problem is the two shapes it accepts and then mishandles.

This is a silent numerical failure, so it is invisible to any test that only checks for exceptions. It looks like the same family as #13699 (TND + `sparse_mode=0` + no mask returning a uniform average), though the trigger here is different. I ran into it while looking for a way to carry a decode sequence length on the device instead of in `actual_seq_lengths_kv` (context in #13058); the report is independent of that.

### Reproduce

Paged decode geometry, one layer, one query token, bf16, no graphs, no vLLM — pure `torch_npu`.

```
ASCEND_RT_VISIBLE_DEVICES=0 python fia_mask_shape_repro.py
```

<details>
<summary>fia_mask_shape_repro.py</summary>

```python
"""Minimal repro: npu_fused_infer_attention_score silently accepts under-ranked
mask / bias tensors under sparse_mode=0 and returns wrong results.

Paged decode geometry, one layer, one query token, bf16, no graphs, no vLLM.
Four ways to tell the operator "only the first L of CAP KV slots are real":

  A  atten_mask bool [1, CAP]           -> operator raises (fine: it says no)
  B  atten_mask bool [1, 1, 1, CAP]     -> accepted; correct at short L, wrong at long L
  C  pse_shift       [1, 1, 1, CAP]     -> accepted; wrong at every L
  D  pse_shift       [1, N, 1, CAP]     -> accepted; correct at every L  (control)

B and C produce no error and no warning. Only D is usable.

Usage:  ASCEND_RT_VISIBLE_DEVICES=<n> python fia_mask_shape_repro.py
"""

import torch
import torch_npu

DEV = "npu:0"
torch.npu.set_device(DEV)
torch.manual_seed(0)

NH, HD, BS = 12, 64, 128
CAP = 4096
NUM_BLOCKS = CAP // BS
DT = torch.bfloat16
SCALE = HD**-0.5
LENGTHS = (137, 600, 1000, 1500, 3000, 4095)
TOL = 4e-3  # bf16 rounding against an fp32 reference is ~2e-3 at these shapes

k_cache = torch.randn(NUM_BLOCKS, BS, NH, HD, dtype=DT, device=DEV).view(NUM_BLOCKS, BS, NH * HD)
v_cache = torch.randn(NUM_BLOCKS, BS, NH, HD, dtype=DT, device=DEV).view(NUM_BLOCKS, BS, NH * HD)
block_table = torch.arange(NUM_BLOCKS, dtype=torch.int32, device=DEV).view(1, NUM_BLOCKS)
q = torch.randn(1, NH, HD, dtype=DT, device=DEV)

arange = torch.arange(CAP, dtype=torch.int32, device=DEV)


def reference(L: int) -> torch.Tensor:
    k = k_cache.view(-1, NH, HD)[:L].float()
    v = v_cache.view(-1, NH, HD)[:L].float()
    s = torch.einsum("tnd,snd->nts", q.float(), k) * SCALE
    return torch.einsum("nts,snd->tnd", s.softmax(-1), v)


def run(L: int, *, atten_mask=None, pse_shift=None) -> torch.Tensor:
    out = torch_npu.npu_fused_infer_attention_score(
        query=q, key=k_cache, value=v_cache,
        atten_mask=atten_mask, pse_shift=pse_shift,
        block_table=block_table, input_layout="TND", block_size=BS,
        actual_seq_lengths=[1], actual_seq_lengths_kv=[CAP],
        num_key_value_heads=NH, num_heads=NH, scale=SCALE, sparse_mode=0,
        pre_tokens=2147483647, next_tokens=0,
    )
    return out[0] if isinstance(out, (tuple, list)) else out


def bool_mask(L: int, shape) -> torch.Tensor:
    """True past the end of the sequence, i.e. 'do not attend here'."""
    return (arange >= L).view(shape)


def bias(L: int, shape) -> torch.Tensor:
    """0 inside the sequence, -inf past its end."""
    return torch.where(arange >= L, float("-inf"), 0.0).to(DT).view(shape)


VARIANTS = (
    ("A atten_mask bool [1, CAP]", lambda L: {"atten_mask": bool_mask(L, (1, CAP))}),
    ("B atten_mask bool [1,1,1,CAP]", lambda L: {"atten_mask": bool_mask(L, (1, 1, 1, CAP))}),
    ("C pse_shift      [1,1,1,CAP]", lambda L: {"pse_shift": bias(L, (1, 1, 1, CAP))}),
    ("D pse_shift      [1,N,1,CAP]", lambda L: {"pse_shift": bias(L, (1, 1, 1, CAP)).expand(1, NH, 1, CAP).contiguous()}),
)

print(f"torch {torch.__version__}  torch_npu {torch_npu.__version__}")
print(f"geometry: 1 query token, {NH} heads, head_dim {HD}, block_size {BS}, capacity {CAP}, {DT}")
print(f"tolerance vs fp32 reference: {TOL:g}\n")

header = "variant".ljust(32) + "".join(f"L={L}".rjust(12) for L in LENGTHS)
print(header)
print("-" * len(header))

for label, make in VARIANTS:
    cells = []
    for L in LENGTHS:
        try:
            got = run(L, **make(L))
            torch.npu.synchronize()
            err = (got.float() - reference(L)).abs().max().item()
            cells.append(f"{err:.2e}" + ("" if err < TOL else "!"))
        except Exception as exc:  # noqa: BLE001 - the point is to report whatever it is
            first = str(exc).strip().split("\n")[0]
            cells.append("RAISED" if len(first) > 40 else first[:11])
    print(label.ljust(32) + "".join(c.rjust(12) for c in cells))

print("\n'!' marks a result that is wrong but was returned without any error.")

# Per-head error: if the under-ranked tensor were broadcast over the head
# dimension as the shape implies, every head would be equally right.
print("\nper-head max_abs_err (L=600), to show the failure is not uniform across heads")
print("variant".ljust(32) + "".join(f"h{h}".rjust(9) for h in range(NH)))
for label, make in VARIANTS[1:]:
    got = run(600, **make(600))
    torch.npu.synchronize()
    per_head = (got.float() - reference(600)).abs().amax(dim=(0, 2))
    print(label.ljust(32) + "".join(f"{e:.1e}".rjust(9) for e in per_head.tolist()))

# Where does the bool-mask variant stop being correct?
print("\nB atten_mask bool [1,1,1,CAP]: finer sweep for the boundary")
row = []
for L in (137, 200, 256, 300, 384, 400, 512, 513, 600):
    got = run(L, atten_mask=bool_mask(L, (1, 1, 1, CAP)))
    torch.npu.synchronize()
    err = (got.float() - reference(L)).abs().max().item()
    row.append((L, err))
print("  " + "  ".join(f"L={L}:{e:.1e}{'' if e < TOL else '!'}" for L, e in row))

for label, make in VARIANTS:
    try:
        run(LENGTHS[0], **make(LENGTHS[0]))
    except Exception as exc:  # noqa: BLE001
        print(f"\n{label} raises:\n  {str(exc).strip().splitlines()[0]}")
```

</details>

### Result

Bit-identical on two separate Atlas A3 / 910C machines:

```
torch 2.10.0+cpu  torch_npu 2.10.0
geometry: 1 query token, 12 heads, head_dim 64, block_size 128, capacity 4096, torch.bfloat16
tolerance vs fp32 reference: 0.004

variant                                L=137       L=600      L=1000      L=1500      L=3000      L=4095
--------------------------------------------------------------------------------------------------------
A atten_mask bool [1, CAP]            RAISED      RAISED      RAISED      RAISED      RAISED      RAISED
B atten_mask bool [1,1,1,CAP]       1.03e-03   1.65e-01!   1.90e-01!   1.93e-01!   2.38e-01!   2.46e-01!
C pse_shift      [1,1,1,CAP]       3.81e-01!   2.17e-01!   1.58e-01!   1.43e-01!   5.96e-02!    1.04e-03
D pse_shift      [1,N,1,CAP]        1.03e-03    5.26e-04    6.16e-04    3.74e-04    2.88e-04    2.03e-04

'!' marks a result that is wrong but was returned without any error.
```

The boundary, from the finer sweep in the same script:

```
B atten_mask bool [1,1,1,CAP]: finer sweep for the boundary
  L=137:1.0e-03  L=200:8.8e-04  L=256:1.0e-03  L=300:6.8e-04  L=384:5.9e-04
  L=400:6.5e-04  L=512:5.3e-04  L=513:1.1e-02!  L=600:1.7e-01!
```

Per-head errors at `L=600` are uniformly wrong rather than "head 0 correct, others garbage", so this is not simply a missing broadcast over the head dimension:

```
variant                                h0       h1       h2       h3       h4       h5       h6       h7       h8       h9      h10      h11
B atten_mask bool [1,1,1,CAP]     7.1e-02  1.7e-01  6.9e-02  7.9e-02  9.9e-02  5.1e-02  6.8e-02  6.2e-02  5.9e-02  8.0e-02  8.3e-02  8.1e-02
C pse_shift      [1,1,1,CAP]      1.4e-01  1.6e-01  2.1e-01  1.9e-01  1.6e-01  1.3e-01  1.3e-01  1.4e-01  1.6e-01  1.8e-01  2.2e-01  1.4e-01
D pse_shift      [1,N,1,CAP]      4.5e-04  4.9e-04  4.9e-04  2.6e-04  5.3e-04  3.1e-04  3.6e-04  3.1e-04  4.5e-04  3.7e-04  4.4e-04  4.9e-04
```

Two details that may help localise it:

- **B is correct only when `L <= 512`** — i.e. when the mask's effect is confined to the first 512 positions.
- **C is correct only at `L=4095`** — i.e. when the bias masks almost nothing. Both are consistent with the tensor being consulted for a bounded prefix and ignored beyond it, rather than being read as garbage.

### Expected

Either honour `atten_mask` / `pse_shift` over the whole KV length at these shapes, or reject them the way the rank-2 `[1, S]` mask is already rejected. Silently computing a wrong attention output is the worst of the three.

### Environment

- Atlas A3 / 910C, `npu-smi` reports `Ascend910` with `NPU Name: 9382`
- CANN 9.0.0
- torch 2.10.0, torch_npu 2.10.0, Python 3.12.13, aarch64
- Reproduced on two machines with identical output; no vLLM or vllm-ascend involved in the repro
